doc/en/man/duende.8
.\" Do *not* edit this file; it was automatically generated by ej2man
.\" Look for a name.ej file with the same name as this filename
.\"
.\" Process this file with the following
.\" nroff -man -Tutf8 maradns.8 | tr '\020' ' '
.\"
.\" Last updated Tue Sep  9 06:40:20 2008
.\"
.TH DUENDE 8 "duende" "January 2003" "duende"
.\" We don't want hyphenation (it's too ugly)
.\" We also disable justification when using nroff
.\" Due to the way the -mandoc macro works, this needs to be placed
.\" after the .TH heading
.hy 0
.if n .na
.\"
.\" We need the following stuff so that we can have single quotes
.\" In both groff and other UNIX *roff processors
.if \n(.g .mso www.tmac
.ds aq \(aq
.if !\n(.g .if '\(aq'' .ds aq \'

.SH "NAME"
.PP
duende \- run a child process as a daemon
.SH "DESCRIPTION"
.PP
.B "duende"
makes a given child process a daemon. The standard output
and standard error
of the child process is logged via syslog() with a priority of
LOG_ALERT.
.SH "USAGE"
.PP
.B "duende"
.RB [ options ]
child_process [ arguments passed onto child ]
.SH "DETAILS"
.PP
When
.B "duende"
is invoked, it spawns two processes. In addition to
spawning the daemonized child process,
.B "duende"
also spawns a process
which reads and logs the standard output of the daemonized process. The
parent process stays alive so as to monitor the daemonized process.
.PP
.B "duende"
requires a blank directory named /etc/maradns/logger
to run.
.PP
Should the parent duende process a
HUP signal,
.B "duende"
will restart the child process. Should the
daemonized or logging process received an untrapped HUP signal or exit
with an exit
code of 8,
.B "duende"
will restart the process. Should the daemonized
or logging
process exit for any other reason,
.B "duende"
will send the logger
process a TERM signal and exit.
Should the duende
parent process receive a TERM or INT signal,
.B "duende"
sends all of its
children TERM signals, then exits.
.PP
The duende process must be started as the superuser; this is because
Duende\(aqs intended child processes (maradns and zoneserver) need to
bind
to privileged ports, and because
duende uses a setuid() call to change the user ID of the logging
process
to the user with ID 66.
.SH "OPTIONS"
.PP
Most of the above behaviour can be configured by command line arguments.
.TP 4
.BI -c | --chroot " directory"
The log helper process will change its directory and root to this location.
This defaults to
.B /etc/maradns/logger "."
.TP 4
.BI -i | --ident " string"
The log helper process will use this string to identify the process when
sending messages to syslog.
This defaults to
.B log-help "."
.TP 4
.BI -p | --pid " filename"
If this argument is present, the duende parent process will store its pid in
this file.
.TP 4
.BI -u | --uid " integer"
The log helper process will drop its user id to this. This value defaults to
66.
.TP 4
.BI -g | --gid " integer"
The log helper process will drop its group id to this. This value defaults to
66.
.TP 4
.BI -r | --restart_on " integer"
The log helper process will restart the child process if it exists with this
status code. This defaults to 8.
.SH "LOGGING"
.PP
.B "duende"
uses the syslog() facility to log the standard output of the
program that it invokes. The name of the program (in other words, the
"ident" given to openlog()) defaults to
.BR "log-helper"
but can be overridden with the
.RB "--ident"
argument.
All messages created by the child process are sent
to syslog() with a priority of LOG_ALERT and a "facility"
of LOG_DAEMON (daemon.info in /etc/syslog.conf).
Should duende itself encounter an error, it will send
messages to syslog() with a priority of LOG_ALERT.
.PP
Also, the directory /etc/maradns/logger, while used by duende, is not
used
to store any log messages. That is unless, for some reason, one
configures syslog to store messages there.
.SH "EXAMPLES"
.PP
Using duende to start maradns, where the mararc file is /etc/mararc.2

.B duende maradns
.RB -f
.IR /etc/mararc.2

Using duende to start zoneserver, where the mararc file is
/etc/mararc.4

.B duende zoneserver
.RB -f
.IR /etc/mararc.4

.SH "BUGS"
.PP
.B "Duende"
assumes that all of its children are well-behaved, eating
their vegetables, going to bed when told, and terminating
when receiving a TERM signal.
.SH "SEE ALSO"
.PP
.BR "maradns(8)" ","
.B "syslog(3)"

.br
http://www.maradns.org
.SH "LEGAL DISCLAIMER"
.PP
THIS SOFTWARE IS PROVIDED BY THE AUTHORS \(aq\(aqAS IS\(aq\(aq AND ANY
EXPRESS
OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.SH "AUTHOR"
.PP
Duende and this man page are written by Sam Trenholme. D Richard Felker
III provided some invaluable assistance with the piping code which
.B "duende"
uses.



debian/changelog
maradns (2.0.04-1) experimental; urgency=low

  TODO:
  - pbuilder (NET)
  - Release to experimental (NET)

  QUESTIONS:
  - Could not get zoneserver to listen on ::1
  - deadwood: If I set dns_port it gives a fatal erorr despite docs

  * Allowed watch file to pick up 2.x releases
  * New upstream release
  * Refreshed patches
  * Added override to stop deletion of server/MaraDNS.c.orig
  * Added patch to stop corruption of deadwood source code
  * Added public domain stanza trying to give useful information about
    the public domain status (or otherwise) of certain files.
  * Updated doc-base index file
  * Split out documentation into separate package as it is written from
    an upstream point of view
  * Added askmara-tcp tool to maradns-zoneserver package
  * Added experimental maradns-deadwood package (Closes: #612229)
  * Turned on compilation for IPv6 (Closes: #477787)

 -- Nicholas Bamber <nicholas@periapt.co.uk>  Fri, 06 Jan 2012 08:47:03 +0000

maradns (1.4.09-1) unstable; urgency=medium

  * Updated TODO.Debian
  * Added askmara-tcp to maradns-zoneserver package
  * Made watch file handle latest upstream webpage
  * Changed the name of internal variables to not conflict with
    built-in bash variables (Closes: #653419)
  * New upstream release
    - randomized source of entropy (Closes: #653838)

 -- Nicholas Bamber <nicholas@periapt.co.uk>  Sat, 31 Dec 2011 22:01:07 +0000

maradns (1.4.07-1) unstable; urgency=low

  * Made watch file handle recursive lookups
  * New upstream release (Closes: #607739)
  * Added lintian override concerning DEP-5
  * Added README.source concerning #643043
  * Added patch updating upstream changelog
  * Updated TODO.Debian

 -- Nicholas Bamber <nicholas@periapt.co.uk>  Mon, 21 Nov 2011 22:10:46 +0000

maradns (1.4.06-3) unstable; urgency=low

  * Made duende more generic and Debian compatible
    - Added patch to use argp to parse arguments
    - Added support for --pid, --uid, --chroot, --ident, --restart_on,
      and --gid arguments
    - Put log helper process in chroot
    - Write pid of log helper process to syslog
    - Consolidated waitpid calls so to avoid race condition allowing zombies
    - Adjusted section and priority, cf. #632337
    - Added example to show how to use duende generically
    - Added patch to man page documenting changes to duende
  * Rewrote maradns init script to use new duende features
    - Init script now relatively standard apart from use of duende and 
      support for multiple servers
    - Reload action in maradns script now sends signal to duende which
      restarts maradns process (Closes: #484466)
    - Added Recommends clause for maradns-zoneserver
  * Added comment about #621833 in postrm script and ensured complete removal
    of /etc/maradns directory
  * Added maradns-zoneserver package (Closes: #582069)
    - Note that the reload action is not supported for zoneserver
  * Updated python support
  * Updated TODO.Debian

 -- Nicholas Bamber <nicholas@periapt.co.uk>  Mon, 18 Jul 2011 12:57:54 +0100

maradns (1.4.06-2) unstable; urgency=low

  * Remove second copy of bind2csv
  * Added debian/clean file
  * Updated TODO.Debian
  * Raised standards version to 3.9.2 - no changes
  * Added insserv config fragment to document that maradns provides
    $named (Closes: #551359)
  * Added '--no-create-home' option to adduser so that fresh install
    does not warn about ownership of the home directory mismatching
  * Removed pid clean up code from postinst/postrm as it should be unneeded
  * Added duende to a new package
  * Patch documentation to reflect name change for bind2csv2 utility
  * Rewrote init scripts to be more robust and use LSB
    - logger processes are now separate daemons
    - logger and maradns now have wrapper scripts redirecting IO through a
      named pipe
    - All pid and named pipe files are now in /var/run/maradns/
    - Status action now supported
    - Upgrades now work smoothly (Closes: #459339)
    - Added comment about origins of multiple server logic to init script
    - As a consequence of being explcitly started by start-stop-daemon,
      logger now has / as its working directory (Closes: 589117)

 -- Nicholas Bamber <nicholas@periapt.co.uk>  Thu, 02 Jun 2011 18:33:40 +0100

maradns (1.4.06-1) unstable; urgency=low

  * New maintainer (Closes: #610842)
  * New upstream version
  * Upped standards version to 3.9.1 - no changes
  * Changed to '3.0 (quilt)' format
  * Rewrote debian/rules and added patches to not change source code
  * Resolved python dependencies
  * Refreshed copyright to DEP-5 standard
  * Put 'set -e' in maintainer scripts
  * Added patch to fix spelling mistakes
  * Tightened up installation of binaries, examples, docs and man pages
  * Added patch to improve readability of man pages
  * Registered tutorial HTML with doc-base
  * TODO.Debian: Added plan for future versions 
  * Added extract of bug report about binding to 0.0.0.0
    to doc-base. (Closes: #486497)
  * Install resolvconf-script to examples, cf. #608946
  * Install zoneserver.init to examples, cf. #582069

 -- Nicholas Bamber <nicholas@periapt.co.uk>  Fri, 04 Feb 2011 16:26:50 +0000

maradns (1.4.03-1.1) unstable; urgency=high

  * Non-maintainer upload by the Security Team
  * Fix CVE-2011-0520 (Closes: #610834)

 -- Moritz Muehlenhoff <jmm@debian.org>  Sun, 30 Jan 2011 14:36:36 +0100

maradns (1.4.03-1) unstable; urgency=low

  * New upstream release
  * Fix "1.4.03 fixes security issue" (Closes: #584587)

 -- Kai Hendry <hendry@iki.fi>  Wed, 07 Jul 2010 16:02:16 +0100

maradns (1.4.02-1) unstable; urgency=low

  * New upstream release

 -- Kai Hendry <hendry@iki.fi>  Sun, 24 Jan 2010 12:17:40 +0000

maradns (1.3.07.09-3) unstable; urgency=low

  * Fix "Provide tools included in maradns source package"
   Added tools csv1tocsv2.pl and maraconf.sh (Closes: #507593)

 -- Kai Hendry <hendry@iki.fi>  Mon, 15 Dec 2008 20:57:07 +0000

maradns (1.3.07.09-2) unstable; urgency=high


tools/duende.c
/* Copyright (c) 2003-2008 Sam Trenholme
 *
 * TERMS
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * This software is provided 'as is' with no guarantees of correctness or
 * fitness for purpose.
 */

/* I would like to thank D. Richard Felker III for his invaluable help
   with debugging how the pipes are set up here */

/* This program is a helper application which does the following:

   It starts the maradns process

   It syslog()s maradns' output via a chroot'd unprivledged child process

   If maradns exits with a code of 8, this means that she was given a
   HUP signal.  Restart the MaraDNS process

   If this process gets a HUP signal, restart the MaraDNS process

 */

#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <syslog.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <argp.h>
#include <string.h>
#include "../MaraDns.h"

int got_hup_signal = 0;
int got_term_signal = 0;

/* needed for command line overrides */
struct argument
{
    int duende_uid; /* = DUENDE_LOGGER_UID; */
    int duende_gid; /* = DUENDE_LOGGER_UID; */
    int restart_on_exit;
    const char *duende_chroot;
    const char *duende_pid_file;
    const char *duende_ident;
    int argc;
    char **argv;
};

static const struct argp_option options[] = {
    {"pid",    'p', "FILE", 0, "File used to store pid of duende process"},
    {"uid",    'u', "UID",  0, "Userid log helper should change down to"},
    {"gid",    'g', "GID",  0, "Groupid log helper should change down to"},
    {"chroot", 'c', "DIR",  0, "Directory log helper should change and chroot to"},
    {"ident",  'i', "STR",  0, "How log helper should be identified in syslog"},
    {"restart_on",    'r', "INT",  0, "Exit status on which to restart child process"},
    {0}
};

static error_t
parse_opt(int key, char *arg, struct argp_state *state)
{
    struct argument *arguments = state->input;

    switch(key) {
        case 'c': /* chroot */
            arguments->duende_chroot = arg;
            break;
        case 'i': /* ident */
            arguments->duende_ident = arg;
            break;
        case 'u': /* uid */
            arguments->duende_uid = atoi(arg);
            break;
        case 'r': /* restart on */
            arguments->restart_on_exit = atoi(arg);
            break;
        case 'g': /* gid */
            arguments->duende_gid = atoi(arg);
            break;
        case 'p': /* pid */
            arguments->duende_pid_file = arg;
            break;
        case ARGP_KEY_NO_ARGS:
            argp_usage(state);
        case ARGP_KEY_ARGS:
            arguments->argc = state->argc - state->next;
            arguments->argv = state->argv + state->next;
            break;
        default:
            return ARGP_ERR_UNKNOWN;
    }

    return 0;
}

static char doc[] = "duende -- a program to manage a daemon process, logging its standard output and error to syslog";
static char args_doc[] = "command [arguments...]";

static struct argp argp = {options, parse_opt, args_doc, doc};

/* If we get a HUP signal set the flag so we can restart the MaraDNS child
   process */
void handle_hup() {
    got_hup_signal = 1;
    return;
    }

void handle_term() {
    got_term_signal = 1;
    return;
    }

/* Helper process which syslogs stuff from either MaraDNS' stdout or stderr. */

void log_helper(const struct argument *arguments, int stdout_fd) {
    char out_buf[1024];

    /* We can't use our signal handlers because fgets is blocking */
    signal(SIGTERM,SIG_DFL);
    signal(SIGHUP,SIG_DFL);

    /* Open up the sys log */
    syslog(LOG_ALERT,"%s process starting up", arguments->duende_ident);

    /* Drop all privileges */
    if (chdir(arguments->duende_chroot) != 0) {
        syslog(LOG_ALERT,"Can not enter chroot directory %s", arguments->duende_chroot);
        syslog(LOG_ALERT,"%s","We can not log daemon output");
        printf("Fatal error logging; read syslog\n");
        printf("%s directory required to exist\n",arguments->duende_chroot);
        exit(1);
    }
#if ! (defined __CYGWIN__ || defined QNX)
    if (chroot(arguments->duende_chroot) != 0) {
        syslog(LOG_ALERT,"Can not chroot to directory %s",arguments->duende_chroot);
        printf("Fatal error logging; read syslog\n");
        exit(1);
    }
#endif
    if(setgid(arguments->duende_gid) != 0) {
        syslog(LOG_ALERT,"%s%d","Can not change GID to ", arguments->duende_gid);
        printf("Fatal error logging; read syslog\n");
        exit(1);
    }
    if(setuid(arguments->duende_uid) != 0) {
        syslog(LOG_ALERT,"%s%d","Can not change UID to ", arguments->duende_uid);
        syslog(LOG_ALERT,"%s","We can not log daemon output");
        printf("Fatal error logging; read syslog\n");
        exit(1);
    }


    /* Log both stdout and stderr */
    dup2(stdout_fd,0);
    for(;;) {
        if(fgets(out_buf,1020,stdin) == out_buf)
#ifdef __FreeBSD__
            /* FreeBSD doesn't log daemon.info messages by default; while
             * this can be changed by editing /etc/syslog.conf, it *is*
             * an issue that can cause confusion */
            syslog(LOG_ALERT,"%s",out_buf);
#else /* __FreeBSD__ */
            syslog(LOG_ALERT,"%s",out_buf);
#endif /* __FreeBSD__ */
        }

    syslog(LOG_ALERT,"log_helper process terminating");
    exit(0);

    }


/* If a child exits, see if the child exited with a code of 8 or received
   a HUP signal.  In either of these cases, restart the child daemon and the
   (if needed) logger process).  Otherwise, exit */

void handle_child_exited(int exit_status, pid_t alive, pid_t exited, int restart_code) {
        if(WIFEXITED(exit_status)) { /* Exit with exit code */
            if(WEXITSTATUS(exit_status) != restart_code) { /* Anything but HUP */
                kill(alive,SIGTERM);
                syslog(LOG_ALERT,"Child exited with status %d",exit_status);
                exit(WEXITSTATUS(exit_status));
                }
            }
        if(WIFSIGNALED(exit_status)) { /* Got signal */
            if(WTERMSIG(exit_status) != SIGHUP) {
                syslog(LOG_ALERT,"Child got signal %d",exit_status);
                kill(alive,SIGTERM);
                exit(1);
                }
            }
       /* If you somehow stop the child daemon, we go bye bye */
       if(WIFSTOPPED(exit_status)) {
            syslog(LOG_ALERT,"Child stopped");
            kill(exited,SIGTERM);
            kill(alive,SIGTERM);
            exit(2);
            }
       /* Clean up the system logging */
       syslog(LOG_ALERT,"Cleaning up system logging");
       kill(alive,SIGTERM);
    }

/* The main process forks off the child.  Right now, I will just have
   it fork off the MaraDNS process, hardwired as /usr/sbin/maradns,
   directing her standard output to
   /dev/null.  The revision of this file will correctly handle Mara's
   output
 */

int main(int argc, char **argv) {
    int exit_status;
    int check_pid;
    pid_t pid, log_pid;
    int stream1[2]; /* Used for piping */
    struct argument arguments;

    /* Initialize data structures */
    arguments.duende_chroot = DUENDE_CHROOT_DIR;
    arguments.duende_pid_file = 0;
    arguments.duende_gid = DUENDE_LOGGER_UID;
    arguments.duende_uid = DUENDE_LOGGER_UID;
    arguments.duende_ident = "log-helper";
    arguments.restart_on_exit = 8;
    arguments.argc = 0;
    arguments.argv = 0;

    argp_parse(&argp, argc, argv, ARGP_IN_ORDER, 0, &arguments);

    if (arguments.argv == 0 || arguments.argv[0] == 0) {
        printf("FATAL: Unable to identify command\n");
        exit(1);
    }

    /* Let children know that duende is running */
    if(setenv("DUENDE_IS_RUNNING","1",0) != 0) {
        printf("FATAL: Unable to set environment variable\n");
        exit(1);
    }

    /* The parent immediately exits */
    if(fork() != 0)
        exit(0);

    /* The child becomes a full-fledged daemon */
    setpgid(0,0); /* No longer visible in 'ps' without the 'auxw' argument */

    /* Write our PID to a file if the user so desires us to */
    if(arguments.duende_pid_file) {
        FILE *fp_pid = fopen(arguments.duende_pid_file,"w");
        if (!fp_pid) {
            syslog(LOG_ALERT,"Fatal writing, to PID file, error\n");
            exit(1);
        }
        unsigned int local_pid = getpid();
        fprintf(fp_pid,"%u",local_pid);
        fclose(fp_pid);
    }
    /* Sysadmins expect HUP to reload, so we set that up */
    signal(SIGHUP,handle_hup);
    signal(SIGTERM,handle_term);
    signal(SIGINT,handle_term);

    pid = 0; log_pid = 0;

    openlog(arguments.duende_ident,LOG_PID,LOG_DAEMON);

    for(;;) {
        if (pipe(stream1) != 0) {
            syslog(LOG_ALERT,"Fatal pipe error");
            exit(3);
        }
        pid = fork();
        if (pid == -1) {
            syslog(LOG_ALERT,"Fatal pid error");
            exit(1);
        }
        if (pid == 0) { /* Child; this one execs maradns */
            close(stream1[0]);
            /* Dup the standard output */
            if (dup2(stream1[1],1) != 1) {
                syslog(LOG_ALERT,"Fatal dup2 error 1");
                exit(4);
            }
            /* And the standard error */
            if (dup2(stream1[1],2) != 2) {
                syslog(LOG_ALERT,"Fatal dup2 error 2");
                exit(5);
            }
            execvp(arguments.argv[0],arguments.argv);
            /* OK, not found */
            printf("duende: %s: Command can't run, terminating\n",arguments.argv[0]);
            syslog(LOG_ALERT,"Command can't run, terminating\n");
            exit(1);
        }

        /* Parent */
        close(stream1[1]);
        log_pid = fork();
        if (log_pid == 0) { /* Child to syslog all of MaraDNS' output */
            log_helper(&arguments, stream1[0]);
            syslog(LOG_ALERT,"log_helper finished, terminating\n");
            exit(1);
        }
        for(;;) {
            /* If we got a HUP signal, send it to the child */
            if (got_hup_signal == 1) {
                kill(pid,SIGHUP);
                got_hup_signal = 0;
            }
            /* If we got a TERM or INT signal, send it to the children
               then exit ourselves */
            else if (got_term_signal == 1) {
                /* XXX: make sure term really stops the children */
                kill(pid,SIGTERM);
                kill(log_pid,SIGTERM);
                syslog(LOG_ALERT,"got term signal, terminating\n");
                exit(0);
            }
            sleep(1);
            int wait_pid = waitpid(-1, &exit_status, WNOHANG);
            if (wait_pid == pid) { /* If child ended */
                handle_child_exited(exit_status,log_pid,pid,arguments.restart_on_exit);
                close(stream1[0]);
                break; /* Out of the inner loop; re-start Mara */
            }
            /* If logger terminated */
            else if (wait_pid == log_pid) {
                handle_child_exited(exit_status,pid,log_pid,arguments.restart_on_exit);
                close(stream1[0]);
                break; /* Out of the inner loop; re-start Mara */
            }
            else if (wait_pid > 0) {
                syslog(LOG_ALERT, "unexpected child reaped: %i", wait_pid);
            }
        }
    }
}

